File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        // 그래프 생성
        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        int[] inDegree = new int[numCourses]; //진입차수로 선수과목의 수

        // 그래프 초기화
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }

        // 각 과목의 선수과목과 in-degree 설정
        for (int[] prerequisite : prerequisites) {
            int course = prerequisite[0];
            int preCourse = prerequisite[1];
            graph.get(preCourse).add(course);
            inDegree[course]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        // in-degree가 0인 과목을 큐에 추가
        for (int i = 0; i < numCourses; i++) {
            if (inDegree[i] == 0) {
                queue.add(i);
            }
        }

        // 위상 정렬 수행
        while (!queue.isEmpty()) {
            int currentCourse = queue.poll();
            // 현재 과목의 선수과목들의 in-degree 감소
            for (int nextCourse : graph.get(currentCourse)) {
                inDegree[nextCourse]--;
                // in-degree가 0이 되면 큐에 추가
                if (inDegree[nextCourse] == 0) {
                    queue.add(nextCourse);
                }
            }
        }

        // 모든 과목의 in-degree가 0인지 확인
        for (int degree : inDegree) {
            if (degree > 0) {
                return false;
            }
        }

        return true; // 모든 과목을 수강할 수 있는 경우
    }
}

File: 0207-course-schedule/0207-course-schedule.c
#include <stdio.h>
#include <stdlib.h>

#define MAX_COURSES 2000

// 그래프 노드 정의
typedef struct Node {
    int course;
    struct Node* next;
} Node;

// 그래프 정의
typedef struct {
    Node* adjList[MAX_COURSES];
    int inDegree[MAX_COURSES];
} Graph;

// 큐 정의
typedef struct {
    int items[MAX_COURSES];
    int front;
    int rear;
} Queue;

// 큐 초기화
void initQueue(Queue* q) {
    q->front = -1;
    q->rear = -1;
}

// 큐가 비어있는지 확인
int isEmpty(Queue* q) {
    return q->front == -1;
}

// 큐에 원소 추가
void enqueue(Queue* q, int item) {
    if (isEmpty(q)) {
        q->front = 0;
    }
    q->rear++;
    q->items[q->rear] = item;
}

// 큐에서 원소 제거
int dequeue(Queue* q) {
    if (isEmpty(q)) {
        exit(EXIT_FAILURE);
    }

    int item = q->items[q->front];
    q->front++;
    if (q->front > q->rear) {
        initQueue(q);
    }
    return item;
}

// 그래프 초기화
void initGraph(Graph* graph, int numCourses) {
    for (int i = 0; i < numCourses; i++) {
        graph->adjList[i] = NULL;
        graph->inDegree[i] = 0;
    }
}

// 그래프에 간선 추가
void addEdge(Graph* graph, int course, int prerequisite) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    if (newNode == NULL) {
        // 메모리 할당 실패 처리
        exit(EXIT_FAILURE);
    }
    newNode->course = course;
    newNode->next = NULL;  // 초기화 추가

    // 그래프의 연결리스트에 노드 추가
    newNode->next = graph->adjList[prerequisite];
    graph->adjList[prerequisite] = newNode;

    graph->inDegree[course]++;
}

// 위상 정렬 수행
bool topologicalSort(Graph* graph, int numCourses) {
    Queue q;
    initQueue(&q);

    // 진입 차수가 0인 노드를 큐에 추가
    for (int i = 0; i < numCourses; i++) {
        if (graph->inDegree[i] == 0) {
            enqueue(&q, i);
        }
    }

    // 위상 정렬 수행
    while (!isEmpty(&q)) {
        int currentCourse = dequeue(&q);

        // 현재 과목과 연결된 노드의 진입 차수 감소
        Node* neighbor = graph->adjList[currentCourse];
        while (neighbor != NULL) {
            graph->inDegree[neighbor->course]--;
            if (graph->inDegree[neighbor->course] == 0) {
                enqueue(&q, neighbor->course);
            }
            neighbor = neighbor->next;
        }
    }

    // 모든 과목의 진입 차수가 0인지 확인
    for (int i = 0; i < numCourses; i++) {
        if (graph->inDegree[i] > 0) {
            return false; // 순환이 발견되면 false 반환
        }
    }

    return true; // 순환이 없으면 true 반환
}

bool canFinish(int numCourses, int** prerequisites, int prerequisitesSize, int* prerequisitesColSize) {
    Graph graph;
    initGraph(&graph, numCourses);

    for (int i = 0; i < prerequisitesSize; i++) {
        int currentCourse = prerequisites[i][0]; // 현재 과목
        for (int j = 1; j < prerequisitesColSize[i]; j++) {
            int prerequisiteCourse = prerequisites[i][j]; // 선수과목
            addEdge(&graph, currentCourse, prerequisiteCourse);
        }
    }

    return topologicalSort(&graph, numCourses);
}